src/core/algorithms/dd/split/split.cpp

```diff
@@ -0,0 +1,702 @@
+    std::size_t last_index = *indices.rbegin();
+    std::list<df> d = SearchSpace(last_index);
+    indices.pop_back();
+    std::list<df> e = SearchSpace(indices);
+    std::list<df> f;
+    df g = min_max_dif_;
+    for (auto const& first_df : e) {
+        for (auto const& second_df : d) {
+            for (std::size_t k = 0; k < num_columns_; k++) {
+                g[k] = {std::max(first_df[k].first, second_df[k].first),
+                        std::min(first_df[k].second, second_df[k].second)};
+            }
+            if (IsFeasible(g))
+                f.push_back(g);
+            else
+                break;
+        }
+    }
+    return f;
+}
+
+bool Split::Subsume(df const& df1, df const& df2) {
+    for (std::size_t i = 0; i < num_columns_; i++) {
+        if (df2[i].first < df1[i].first || df2[i].second > df1[i].second) return false;
+    }
+    return true;
+}
+
+std::list<dd> Split::NegativePruningReduce(df const& rhs, std::list<df> const& search,
+                                           unsigned& cnt) {
+    if (!search.size()) return {};
+
+    std::list<dd> dds;
+    df last_df = *search.rbegin();
+
+    cnt++;
+    if (!VerifyDD({last_df, rhs})) {
+        std::list<df> remainder;
+        for (auto const& search_df : search) {
+            if (search_df != last_df) {
+                if (!Subsume(search_df, last_df)) {
+                    remainder.push_back(search_df);
+                }
+            }
+        }
+        return NegativePruningReduce(rhs, remainder, cnt);
+    }
+
+    std::list<df> prune;
+    std::list<df> remainder;
+    for (auto const& search_df : search) {
+        if (search_df != last_df) {
+            if (Subsume(search_df, last_df)) {
+                prune.push_back(search_df);
+            } else {
+                remainder.push_back(search_df);
+            }
+        }
+    }
+
+    dds = NegativePruningReduce(rhs, prune, cnt);
+    if (!dds.size()) dds.push_back({last_df, rhs});
+    std::size_t cur_size = dds.size();
+    std::list<dd> remaining_dds = NegativePruningReduce(rhs, remainder, cnt);
+
+    for (auto const& remaining_dd : remaining_dds) {
+        bool h = true;
+        std::size_t index = 0;
+        for (auto const& pruning_dd : dds) {
+            if (Subsume(pruning_dd.lhs, remaining_dd.lhs)) {
+                h = false;
+                break;
+            }
+            index++;
+            if (index == cur_size) break;
+        }
+        if (h) dds.push_back(remaining_dd);
+    }
+
+    return dds;
+}
+
+std::list<dd> Split::HybridPruningReduce(df const& rhs, std::list<df> const& search,
+                                         unsigned& cnt) {
+    if (!search.size()) return {};
+
+    std::list<dd> dds;
+    df first_df = *search.begin();
+    df last_df = *search.rbegin();
+
+    cnt++;
+    if (VerifyDD({first_df, rhs})) {
+        dds.push_back({first_df, rhs});
+        std::list<df> remainder;
+        for (auto const& search_df : search) {
```

Comment:
We have this pattern like in many places, Could we warp it into function and name like `DoPositivePruning`?